### PR TITLE
[WIP] New `show` methods in `QuadForm`

### PR DESCRIPTION
**Pull request by StevellM**
Following the new Oscar printing guideline, here is a first round of suggestions for the new `show` methods in the context of `QuadForm/`.

I am not strict on the decisions, this is why I called these suggestions. The branch is stable so anyone who wants to play around with the new show methods can just run Hecke on it.

Some printings rely on compact/non-compact printings for prime ideals, maximal orders and number fields. So up to now, not all the printings are completely compact, or they do not completely follow the new guideline.

In the same way, the printing of `TorQuadModuleMor` and `AbstractSpaceMor` depends on the printing system of `HeckeMap`.

@simonbrandhorst @thofma 

Here a overview of what one could see:

<details><summary>Lattice Databases</summary>
<p>

```
julia> ld = lattice_database()
Definite integer lattices of rank <= 40
Author: Gabriele Nebe and Neil Sloane
Source: http://www.math.rwth-aachen.de/~Gabriele.Nebe/LATTICES/index.html#25D
Number of lattices: 657

julia> show(stdout, ld)
Nebe-Sloan database of lattices (rank limit = 40)

julia> show(IOContext(stdout, :supercompact => true), ld)
Integer lattices database
```
```
julia> qld = Hecke.quadratic_lattice_database()
Quadratic lattices of rank >= 3 with class number 1 or 2
Author: Markus Kirschmer
Source: http://www.math.rwth-aachen.de/~Markus.Kirschmer/forms/
Version: 0.0.1
Number of lattices: 30250

julia> show(stdout, qld)
Quadratic lattices of rank >= 3 with class number 1 or 2

julia> show(IOContext(stdout, :supercompact => true), qld)
Quadratic lattices database
```
```
julia> hld = Hecke.hermitian_lattice_database()
Hermitian lattices of rank >= 3 with class number 1 or 2
Author: Markus Kirschmer
Source: http://www.math.rwth-aachen.de/~Markus.Kirschmer/forms/
Version: 0.0.1
Number of lattices: 570

julia> show(stdout, hld)
Hermitian lattices of rank >= 3 with class number 1 or 2

julia> show(IOContext(stdout, :supercompact => true), hld)
Hermitian lattices database
```

</p>
</details> 

<details><summary>Binary quadratic forms</summary>
<p>

```
julia> q = binary_quadratic_form(2,3,4)
Binary quadratic form
  over integer ring
with equation 2*x^2 + 3*x*y + 4*y^2

julia> show(stdout, q)
Binary quadratic form over ZZ: 2*x^2 + 3*x*y + 4*y^2

julia> show(IOContext(stdout, :supercompact => true), q)
Binary quadratic form
```

</p>
</details> 

<details><summary>Line enumeration</summary>
<p>

```
julia> P = Hecke.enumerate_lines(GF(3), 7)
Iterator for vector lines in a 7-dimensional vector space
  over finite field of characteristic 3

julia> show(stdout, P)
Iterator for Gr(1, 7) over GF(3)

julia> show(IOContext(stdout, :supercompact => true), P)
Iterator for vector lines
```

</p>
</details> 

<details><summary>Change of scalars</summary>
<p>

```
julia> K, _ = rationals_as_number_field();

julia> f = Hecke.VecSpaceRes(K, 4)
Map of change of scalars
  from vector space of dimension 4 over rational field
  to vector space of dimension 4 over number field of degree 1 over QQ

julia> show(stdout, f)
Map of change of scalars between vector spaces

julia> show(IOContext(stdout, :supercompact => true), f)
Map of change of scalars
```
```
julia> E,b = cyclotomic_field_as_cm_extension(13);

julia> V = hermitian_space(E, 5);

julia> res = restrict_scalars(V, QQ)[2]
Map of change of scalars
  from quadratic space of dimension 60
  to hermitian space of dimension 5

julia> show(stdout, res)
Map of change of scalars between hermitian spaces

julia> show(IOContext(stdout, :supercompact => true), res)
Map of change of scalars
```

</p>
</details> 

<details><summary>Torsion quadratic modules</summary>
<p>

```
julia> L = root_lattice(:E, 7);

julia> qL = discriminant_group(L)
Finite quadratic module
  over integer ring
Abelian group: Z/2
Bilinear value module: Q/Z
Quadratic value module: Q/2Z
Gram matrix quadratic form: 
[3//2]

julia> show(stdout, qL)
Finite quadratic module: Z/2 -> Q/2Z

julia> show(IOContext(stdout, :supercompact => true), qL)
Finite quadratic module
```
```
julia> qL[1]
Element
  of finite quadratic module: Z/2 -> Q/2Z
with component [1]

julia> show(stdout, qL[1])
[1]

julia> show(IOContext(stdout, :supercompact => true), qL[1])
Element of finite quadratic module
```

</p>
</details> 

<details><summary>Local multiplicative groups modulo squares</summary>
<p>

```
julia> E,b = cyclotomic_field_as_cm_extension(13);

julia> K = base_field(E);

julia> OK = maximal_order(K);

julia> p = prime_decomposition(OK, 3)[1][1];

julia> f = Hecke.local_multiplicative_group_modulo_squares(p)[2]
Map for local unit group modulo squares
  from GrpAb: (Z/2)^2
  to maximal real subfield of cyclotomic field of order 13
at the prime ideal <3, (z_13 + 1//z_13)^3 + (z_13 + 1//z_13)^2 + 2*(z_13 + 1//z_13) + 1>

julia> show(stdout, f)
Map for local unit group modulo squares at the prime ideal <3, (z_13 + 1//z_13)^3 + (z_13 + 1//z_13)^2 + 2*(z_13 + 1//z_13) + 1>

julia> show(IOContext(stdout, :supercompact => true), f)
Map for local unit group modulo squares
```

</p>
</details> 

<details><summary>Integer lattices</summary>
<p>

```
julia> L = root_lattice(:E, 7)
Integer lattice of rank 7 and degree 7
with gram matrix
[ 2   -1    0    0    0    0    0]
[-1    2   -1    0    0    0    0]
[ 0   -1    2   -1    0    0   -1]
[ 0    0   -1    2   -1    0    0]
[ 0    0    0   -1    2   -1    0]
[ 0    0    0    0   -1    2    0]
[ 0    0   -1    0    0    0    2]

julia> show(stdout, L)
Integer lattice of rank 7 and degree 7

julia> show(IOContext(stdout, :supercompact => true), L)
Integer lattice
```
```
julia> G = genus(L)
Genus symbol for integer lattices
Signatures: (7, 0, 0)
Local symbol: 
  Local genus symbol at 2: 1^6 2^1_7

julia> show(stdout, G)
Genus symbol: II_(7, 0) 2^1_7

julia> show(IOContext(stdout, :supercompact => true), G)
II_(7, 0) 2^1_7

julia> show(stdout, "text/latex", G)
II_{(7, 0)}2^{1}_{7}
```
```
julia> g = genus(L, 2)
Local genus symbol for integer lattices
Prime: 2
Jordan blocks (val, rank, det, sign, oddity): 
  (0, 6, 7, 0, 0)
  (1, 1, 7, 1, 7)

julia> show(stdout, g)
Local genus symbol at 2: 1^6 2^1_7

julia> show(IOContext(stdout, :supercompact => true), g)
2: (0, 6, 7, 0, 0)(1, 1, 7, 1, 7)

julia> show(stdout, "text/latex", g)
2^{1}_{7}
```

</p>
</details> 

<details><summary>Quadratic spaces</summary>
<p>

```
julia> V = ambient_space(L)
Quadratic space of dimension 3
  over mumber field of degree 1 over QQ
with gram matrix
[4    0    0]
[0   26    0]
[0    0   52]

julia> show(stdout, V)
Quadratic space of dimension 3

julia> show(IOContext(stdout, :supercompact => true), V)
Quadratic space
```
```
julia> p = prime(genus(L).LGS[1]);

julia> Ip = Hecke.isometry_class(V, p)
Local isometry class
  of quadratic space
    over number field with defining polynomial x - 1
      over rational field
Prime ideal: <13, 13>
Invariants: 
  Dimension: 3
  Determinant: 5408
  Hasse invariant: -1

julia> show(stdout, Ip)
Isometry class of quadratic spaces over the 13-adic integers

julia> show(IOContext(stdout, :supercompact => true), Ip)
Local isometry class of quadratic spaces
```
```
julia> I = Hecke.isometry_class(V)
Isometry class
  of quadratic space
    over number field with defining polynomial x - 1
      over rational field
Invariants:
  Dimension: 3
  Determinant: 5408
  Signature tuple: [(3, 0, 0)]
Negative Hasse invariants at: <13, 13>, <2, 2>

julia> show(stdout, I)
Isometry class of quadratic space over number field

julia> show(IOContext(stdout, :supercompact => true), I)
Isometry class of quadratic space
```

</p>
</details> 

<details><summary>Quadratic lattices</summary>
<p>

```
julia> L = lattice(qld, 18)
Quadratic lattice of rank 3 and degree 3
  over maximal order of Number field of degree 1 over QQ
  with basis nf_elem[1]

julia> show(stdout, L)
Quadratic lattice of rank 3 and degree 3

julia> show(IOContext(stdout, :supercompact => true), L)
Quadratic lattice
```
```
julia> p = prime(genus(L).LGS[1]);

julia> J = Hecke.JorDec(L, p)
Jordan decomposition for quadratic lattices
  over maximal order of Number field of degree 1 over QQ
  with basis nf_elem[1]
Prime ideal: <13, 13>
Jordan blocks (scale, rank, determinant class):
  (2, 1, 18928)
  (1, 2, 338//7)

julia> show(stdout, J)
Jordan decomposition for quadratic lattices over the 13-adic integers

julia> show(IOContext(stdout, :supercompact => true), J)
<13, 13>: (1, 2, 18928)(2, 1, 338//7)
```
```
julia> g = genus(L, p)
Local genus symbol for quadratic lattices
  over maximal order of Number field of degree 1 over QQ
  with basis nf_elem[1]
Prime ideal: <13, 13>
Unifomizer: 13
Jordan blocks (scale, rank, determinant class):
  (1, 2, -1)
  (2, 1, 1)

julia> show(stdout, g)
Local genus symbol for quadratic lattices over the 13-adic integers

julia> show(IOContext(stdout, :supercompact => true), g)
(1, 2, -1)(2, 1, 1)
```
```
julia> G = genus(L)
Genus symbol for quadratic lattices
  over maximal order of Number field of degree 1 over QQ
  with basis nf_elem[1]
Signature: 
  Infinite place corresponding to (Complex embedding of number field) => 0
Local symbols: 
  <13, 13> => (1, 2, -1)(2, 1, 1)
  <2, 2> => (0, 2, 2, 1, 3//13, 1)(1, 1, 26//3, 2, 26//3, 1)

julia> show(stdout, G)
Genus symbol for quadratic lattices of rank 3 over maximal order of Number field of degree 1 over QQ with basis nf_elem[1]

julia> show(IOContext(stdout, :supercompact => true), G)
Genus symbol for quadratic lattices
```

</p>
</details> 

<details><summary>Hermitian spaces</summary>
<p>

```
julia> E,b = cyclotomic_field_as_cm_extension(13);

julia> V = hermitian_space(E, 3)
Hermitian space of dimension 3
  over relative number field with defining polynomial t^2 - (z_13 + 1//z_13)*t + 1
    over number field with defining polynomial $^6 + $^5 - 5*$^4 - 4*$^3 + 6*$^2 + 3*$ - 1
      over rational field
with gram matrix
[1   0   0]
[0   1   0]
[0   0   1]

julia> show(stdout, V)
Hermitian space of dimension 3

julia> show(IOContext(stdout, :supercompact => true), V)
Hermitian space
```

</p>
</details> 

<details><summary>Hermitian lattices</summary>
<p>

```
julia> hld = Hecke.hermitian_lattice_database();

julia> H = lattice(hld, 384)
Hermitian lattice of rank 4 and degree 4
  over relative maximal order of relative number field of degree 2 over number field
  with pseudo-basis 
  (1, 1//1 * <1, 1>)
  (b + 1, 1//2 * <1, 1>)

julia> show(stdout, H)
Hermitian lattice of rank 4 and degree 4

julia> show(IOContext(stdout, :supercompact => true), H)
Hermitian lattice
```
```
julia> G = genus(H)
Genus symbol for hermitian lattices
  over relative maximal order of relative number field of degree 2 over number field
  with pseudo-basis
  (1, 1//1 * <1, 1>)
  (b + 1,  1//2 * <1, 1>)
Signature: 
  Infinite place corresponding to (Complex embedding of number field) => 0
Local symbols:
  <2, 2> => (0, 3, +)(2, 1, +)
  <3, 3> => (0, 4, +)

julia> show(stdout, G)
Genus symbol for hermitian lattices of rank 4 over relative maximal order of relative number field  with pseudo-basis 
(1, 1//1 * <1, 1>)
(b + 1, 1//2 * <1, 1>)

julia> show(IOContext(stdout, :supercompact => true), G)
Genus symbol for hermitian lattices
```
```
julia> g = genus(H).LGS[1]
Local genus symbol for hermitian lattices
  over relative maximal order of relative number field of degree 2 over number field 
  with pseudo-basis
  (1, 1//1 * <1, 1>)
   (b + 1, 1//2 * <1, 1>)
Prime ideal: <2, 2>
Jordan blocks (scale, rank, det):
  (0, 3, +)
  (2, 1, +)

julia> show(stdout, g)
Local genus symbol for hermitian lattices over the 2-adic integers

julia> show(IOContext(stdout, :supercompact => true), g)
(0, 3, +)(2, 1, +)
```

</p>
</details>

### CODE REVIEW

**Comment by simonbrandhorst** [comment]
I think it enumerates subspaces of dimension 1 and not affine lines.
```
julia> P = Hecke.enumerate_lines(GF(3), 7)
Iterator for affine lines in a 7-dimensional space
  over Galois field with characteristic 3

julia> show(stdout, P)
Iterator for Gr(1, 7) over Galois field with characteristic 3

julia> show(IOContext(stdout, :supercompact => true), P)
Iterator for affine lines
```
Are they affine? I would have expected them to be subspaces of dimension 1. 

I would prefer the following:
```
julia> qL[1]
Element
  of Finite quadratic module: Z/2 -> Q/2Z
with components [1]

julia> show(stdout, qL[1])
[1]

julia> show(IOContext(stdout, :supercompact => true), qL[1])
Element of finite quadratic module
```

Use the singular for a single space:
```
julia> p = prime(genus(L).LGS[1]);

julia> Ip = Hecke.isometry_class(V, p)
Local isometry class of quadratic space
  over Number field over Rational Field with defining polynomial x - 1
Prime ideal: <13, 13>
Invariants: 
  Dimension: 3
  Determinant: 5408
  Hasse invariant: -1

julia> show(stdout, Ip)
Isometry class of quadratic space over the 13-adic integers

julia> show(IOContext(stdout, :supercompact => true), Ip)
Local isometry class of quadratic space
```

Thank you very much for this PR! It looks nice.

**Comment by StevellM** [comment]
Thanks for the comments!

> Are they affine? I would have expected them to be subspaces of dimension 1.

In fact yes, they are 1-dimensional sub vector spaces... I used affine as it was written above, and I kept it thinking about the distinction with projective lines. I will change that.

I do not understand though
> Use the singular for a single space:

I guess that in an (local) isometry class, there are several spaces. So in the same way I have written "Genus symbol for (bla) lattice**s**", I would expect to put "space" in plural for the definition of such objects. Am I missing something ?

File: src/Assertions.jl [hunk 1]
```diff
@@ -282,7 +282,7 @@ verbosity scopes by calling the function [`add_verbosity_scope`](@ref).
 julia> add_verbosity_scope(:MyScope1);
 
 julia> get_verbosity_level(:MyScope1)
-0
+4
```
Comment (thofma) [inline src/Assertions.jl:285]: This should stay `0`, but you ran the tests too often :)

Comment (StevellM) [reply src/Assertions.jl:285]: Yes indeed, I am currently fixing it, I just noticed that it was wrong when pushing the new commits.

**Comment by StevellM** [comment]
@simonbrandhorst  here is the last update with the decapitalizations and smart indents. I have fixed for printing empty genera and I changed the printing of `TorQuadMod` so that the underlying abelian group is printing in snf (so it doesnt't match exactly with the size of the gram matrix, but we have nicer printing in that way). You can refer to my very comment where I have updated all the printings.

Minor comments: the printings for maximal orders are still not in new format, so it will change once changed. For prime ideals in number fields, the `:supercompact` option is not yet available, but we have `:compact` which I have used for now.

**Comment by simonbrandhorst** [comment]
I would print the gram matrix of a lattice of rank <= 24 in detailed mode. 

```
julia> G = genus(L)
Genus symbol for integer lattices
Signatures: (7, 0, 0)
Local symbol: 
  Local genus symbol at 2: 1^6 2^1_7
```
Does it print the `Local symbols` in plural in case there are several?
Thanks for the update.

**Comment by simonbrandhorst** [comment]
I can live with that.

**Comment by StevellM** [comment]
> I would print the gram matrix of a lattice of rank <= 24 in detailed mode.

For all kind of lattices, or only integer ones ? 

 >Does it print the `Local symbols` in plural in case there are several? Thanks for the update.
  
 If I did not miss anything, everytime there is a possible plural, we have a case distinction in the printing functions.

**Comment by simonbrandhorst** [comment]
Perfect. Printing the gram matrix is useful for `Z`-lattices only. The other ones may not be free and then there is no gram matrix.

**Comment by thofma** [comment]
Yes, only Z-lattices. Thanks for working on this @StevellM.

File: src/QuadForm/Herm/Genus.jl [hunk 2]
```diff
@@ -18,50 +18,64 @@ local_genus_herm_type(E) = HermLocalGenus{typeof(E), ideal_type(order_type(base_
...
-    if length(G) > 0
-      print(IOContext(io, :compact => true), prime(G), ":")
-      print(io, "\n")
-    else
-      print(IOContext(io, :compact => true), prime(G), " of rank zero")
-    end
+  p = prime(G)
+  println(io, "Local genus symbol for hermitian lattices")
+  nio = pretty(io)
+  io = pretty(io)
```
Comment (thofma) [inline src/QuadForm/Herm/Genus.jl:24]: Why is this necessary?

File: src/QuadForm/Quad/ZGenus.jl [hunk 3]
```diff
@@ -2587,12 +2706,31 @@ julia> NS = direct_sum(integer_lattice(:U), rescale(root_lattice(:A, 16), -1))[1
 julia> LK3, iNS, i = embed_in_unimodular(NS, 3, 19);
 
 julia> genus(LK3)
-ZZGenus
-Signature: (3, 19)
-Genus symbol at 2:   1^22
+Genus symbol for integer lattices
+Signatures: (3, 0, 19)
+Local symbol:
```
Comment (thofma) [inline src/QuadForm/Quad/ZGenus.jl:2711]: It seems nothing is printed for the local symbol?

File: src/QuadForm/Herm/Genus.jl (hunk 2 above)
Comment (StevellM) [reply src/QuadForm/Herm/Genus.jl:24]: Glad you bring this up! The `Lowercase()` is global on your `IOContext`. So for everywhere else you use an `IO` on which you have put `Lowercase()`, the `lowercase_first` is applied, resulting to non-expected lowercasing. For instance, I want to lowercase the first letter for the maximal order, and so I call `Lowercase()` for the `nio` in line 25. But for my printing of `Jordan block`, I want to keep the uppercase: if I use the same `nio`, everything is going to be lowercased, this is why I produce a "different" pretty `IO` to avoid that issue. I hope it seems clear.

File: src/QuadForm/Quad/ZGenus.jl (hunk 3 above)
Comment (StevellM) [reply src/QuadForm/Quad/ZGenus.jl:2711]: Ah yes indeed, I will have a look at it.. thanks!

**Review by simonbrandhorst — COMMENTED** [review]
Looks good from my side. Can this be merged @thofma ?
